staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go

```diff
@@ -0,0 +1,312 @@
+
+		allKinds.Insert(curr.Status.AcceptedNames.Kind)
+		allKinds.Insert(curr.Status.AcceptedNames.ListKind)
+	}
+
+	return allResources, allKinds
+}
+
+func (c *NamingConditionController) calculateNames(in *apiextensions.CustomResourceDefinition) (apiextensions.CustomResourceDefinitionNames, apiextensions.CustomResourceDefinitionCondition) {
+	// Get the names that have already been claimed
+	allResources, allKinds := c.getNamesForGroup(in.Spec.Group)
+
+	condition := apiextensions.CustomResourceDefinitionCondition{
+		Type:   apiextensions.NameConflict,
+		Status: apiextensions.ConditionUnknown,
+	}
+
+	requestedNames := in.Spec.Names
+	acceptedNames := in.Status.AcceptedNames
+	newNames := in.Status.AcceptedNames
+
+	// Check each name for mismatches.  If there's a mismatch between spec and status, then try to deconflict.
+	// Continue on errors so that the status is the best match possible
+	if err := checkName(requestedNames.Plural, acceptedNames.Plural, allResources); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "Plural"
+		condition.Message = err.Error()
+	} else {
+		newNames.Plural = requestedNames.Plural
+	}
+	if err := checkName(requestedNames.Singular, acceptedNames.Singular, allResources); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "Singular"
+		condition.Message = err.Error()
+	} else {
+		newNames.Singular = requestedNames.Singular
+	}
+	if !reflect.DeepEqual(requestedNames.ShortNames, acceptedNames.ShortNames) {
+		errs := []error{}
+		existingShortNames := sets.NewString(acceptedNames.ShortNames...)
+		for _, shortName := range requestedNames.ShortNames {
+			// if the shortname is already ours, then we're fine
+			if existingShortNames.Has(shortName) {
+				continue
+			}
+			if err := checkName(shortName, "", allResources); err != nil {
+				errs = append(errs, err)
+			}
+
+		}
+		if err := utilerrors.NewAggregate(errs); err != nil {
+			condition.Status = apiextensions.ConditionTrue
+			condition.Reason = "ShortNames"
+			condition.Message = err.Error()
+		} else {
+			newNames.ShortNames = requestedNames.ShortNames
+		}
+	}
+
+	if err := checkName(requestedNames.Kind, acceptedNames.Kind, allKinds); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "Kind"
+		condition.Message = err.Error()
+	} else {
+		newNames.Kind = requestedNames.Kind
+	}
+	if err := checkName(requestedNames.ListKind, acceptedNames.ListKind, allKinds); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "ListKind"
+		condition.Message = err.Error()
+	} else {
+		newNames.ListKind = requestedNames.ListKind
+	}
+
+	// if we haven't changed the condition, then our names must be good.
+	if condition.Status == apiextensions.ConditionUnknown {
+		condition.Status = apiextensions.ConditionFalse
+		condition.Reason = "Passed"
+		condition.Message = "no conflicts found"
+	}
+
+	return newNames, condition
```

Comment:
Would the semantics be much simpler if we either accepts all requested names or none? Now it can happen that the plural changes, but the singular has a conflicts. Don't like that very much.